core/src/smartcontracts/isi/permissions/roles.rs

```diff
@@ -172,3 +203,79 @@ pub fn unpack_if_role_revoke(
 ) -> eyre::Result<Vec<Instruction>> {
     unpack!(instruction, wsv, Instruction::Revoke => RevokeBox)
 }
+
+macro_rules! impl_extract_specialized_token {
+    (<$isi_type:ty>, $isi:ident, $wsv:ident) => {{
+        let value = $isi
+            .object
+            .evaluate($wsv, &Context::new())
+            .map_err(|e| e.to_string())?;
+
+        match value {
+            Value::Id(IdBox::RoleId(role_id)) => {
+                let role = $wsv
+                    .roles()
+                    .get(&role_id)
+                    .ok_or_else(|| format!("Role with id `{role_id}` not found"))?;
+                let specialized_token = role
+                    .permissions()
+                    .find_map(|permission| T::try_from(permission.clone()).ok())
+                    .ok_or_else(|| {
+                        format!(
+                            "Role {} doesn't contain requested permission token",
+                            role.value()
+                        )
+                    })?;
+
+                Ok(specialized_token)
+            }
+            Value::PermissionToken(permission_token) => {
+                let specialized_token: T = permission_token
+                    .try_into()
+                    .map_err(|e: PredefinedTokenConversionError| e.to_string())?;
+
+                Ok(specialized_token)
+            }
+            _ => Err(format!(
+                "Provided `{}` instruction contains unsupported object type",
+                stringify!($isi_type)
+            )),
+        }
+    }};
+}
+
+/// Extracts specialized token from [`GrantBox`]
+///
+/// # Errors
+/// - Cannot evaluate `instruction`
+/// - `instruction` doesn't evaluate to [`RoleId`] or [`PermissionToken`]
+/// - There is no such role
+/// - Role doesn't contain requested specialized token
```

Comment:
```suggestion
/// - `instruction` doesn't evaluate to [`RoleId`] or [`PermissionToken`].
/// - There is no such role.
/// - Role doesn't contain requested specialized token.
```

Reply:
As this is the enumeration I'd like to use `;`